### ISSUE TITLE
OSDI 25 | Mirage: A Multi-Level Superoptimizer for Tensor Programs

### ISSUE THREAD

**Issue by woshipapa** [AI-Infra-Team/awesome-papers#190]
Mirage: A Multi-Level Superoptimizer for Tensor Programs

[Paper](https://www.usenix.org/system/files/osdi25-wu-mengdi.pdf)
[Code](https://github.com/mirage-project/mirage)

## **Motivation**

As DNN execution on GPUs increasingly depends on high-performance tensor program optimization, existing frameworks and automated optimizers still face a fundamental limitation: they usually optimize only the **composition or scheduling of existing operators/kernels**, and cannot automatically discover new custom kernels such as FlashAttention, which require coordinated optimization across multiple GPU hierarchy levels.

Existing approaches broadly fall into two categories. The first category includes **schedule-based optimizers** such as Halide, TVM, and Ansor. These methods fix the algorithm and search for schedules such as tiling, parallelization, and memory placement. The second category includes **algebraic optimizers** such as TASO, PET, Grappler, and Tensat. These methods exploit mathematical equivalence to rewrite operator-level graphs, for example through operator fusion, convolution-to-matmul transformation, or transformations based on associativity, commutativity, and distributivity.

However, both categories share a common assumption: programmers must provide a fixed set of available operators or kernel implementations. These systems can reorder existing computations or tune how existing kernels execute, but they struggle to introduce entirely new custom kernels.

The paper’s core question is:

**How can a system automatically discover high-performance tensor program optimizations that jointly require algebraic transformations, schedule transformations, and new custom kernel generation?**

This motivation is built on four observations:

**Existing frameworks rely on manual rules and expert-written kernels.**  
Systems such as PyTorch and TensorFlow usually map tensor programs to cuBLAS, cuDNN, or hand-written GPU kernels through manually designed rules. This approach requires substantial engineering effort and can miss more sophisticated optimization opportunities.

**Schedule-based optimization can change “how to compute,” but not “what to compute.”**  
Systems such as TVM and Ansor separate algorithm from schedule and mainly search for efficient ways to execute a given algorithm on the target GPU. However, many linear algebra computations in DNNs have a wide range of mathematically equivalent formulations. If the algorithm is fixed, many potential optimizations are excluded from the search space.

**Algebraic optimizers rewrite existing operator graphs, but remain limited by existing kernels.**  
Systems such as TASO and PET can perform operator fusion and algebraic rewrites, but they still search primarily at the kernel-graph level and require each operator to have an existing implementation. As a result, they struggle to discover new implementations that reorganize computation inside kernels across blocks and threads.

<img width="524" height="339" alt="Image" src="https://github.com/user-attachments/assets/bf559dbe-7fa5-4cec-ab06-d634694ca607" />

<img width="536" height="696" alt="Image" src="https://github.com/user-attachments/assets/a51aba7f-e4ea-4848-b10e-96e5a20884bc" />

**The strongest GPU optimizations often require coordination across the kernel, thread-block, and thread levels.**  
FlashAttention is a representative example. It does not merely fuse operators or tune a given schedule; it reorders the mathematical computation, reorganizes kernel structure, changes block/thread-level parallelization, and exploits shared memory and register files to reduce device-memory traffic. Such optimizations have traditionally required expert-written custom kernels and are outside the search space of existing automated optimizers.

Therefore, the paper argues that high-performance tensor program optimization cannot stop at operator-graph rewriting or single-kernel scheduling. It requires a new representation that can uniformly express computation at the **kernel level, thread-block level, and thread level**.

## **Solution**

To address this problem, the paper proposes **Mirage**, the first **multi-level superoptimizer** for tensor programs. Its central idea is **µGraph**, a unified representation of computation across the GPU hierarchy: kernel, thread block, and thread. This allows Mirage to jointly search over algebraic transformations, schedule transformations, and new custom kernel generation.

The overall pipeline is:

<img width="530" height="659" alt="Image" src="https://github.com/user-attachments/assets/f8199015-6413-4eac-8d1b-7fedfd4f5c1a" />

```text
Input Tensor Program
    ↓
Program Partitioning
    ↓
Expression-Guided µGraph Generator
    ↓
Probabilistic Equivalence Verifier
    ↓
µGraph Optimizer
    ↓
Optimized Tensor Program
```

**Program Partitioning: split the large program into searchable and verifiable LAX subprograms.**  
Mirage first partitions the input tensor program into multiple LAX subprograms. The LAX fragment includes multilinear operators, division, and limited exponentiation, covering common DNN computations such as matrix multiplication, convolution, normalization, and activation. This reduces the search space while also providing the mathematical foundation for probabilistic equivalence verification.

**µGraph: uniformly represent the kernel, block, and thread levels of the GPU hierarchy.**  
µGraph is the core abstraction in Mirage. A kernel graph represents computation among GPU kernels/operators, with tensors stored in device memory. A block graph represents computation inside a thread block, with intermediate tensors stored in shared memory. A thread graph represents computation inside a single thread, with intermediate values stored in registers. This multi-level representation allows Mirage to express operator fusion, memory hierarchy reuse, block/thread-level tiling, and entirely new custom kernel structures.

**Expression-Guided µGraph Generator: automatically discover candidate µGraphs.**  
For each LAX subprogram, Mirage searches for potentially equivalent µGraphs. Since the search space is much larger than that of traditional superoptimizers, Mirage uses a hybrid strategy: it exhaustively explores graphs up to a certain size at the kernel and block levels, while using rule-based fusion to construct thread-level graphs. This preserves the most performance-critical optimization opportunities while avoiding an explosion in the thread-level search space.

**Abstract Expression Pruning: prune the search space using abstract expressions.**  
Mirage computes an abstract expression for each intermediate tensor to determine whether the current graph prefix can contribute to the target computation. If a prefix’s abstract expression cannot be part of the desired output computation, it is pruned early. This mechanism reduces the search space while avoiding overly aggressive pruning of useful prefixes, and provides a certain optimality guarantee.

**Thread Graph Construction: fuse local computation into register-level execution.**  
For thread-level computation, Mirage avoids exhaustive search and instead uses a transformation-based construction method. For example, it can fuse elementwise operators such as Mul, Sqrt, and Div into a single thread graph, keeping intermediate values in registers instead of writing them back to shared memory.

**Probabilistic Equivalence Verifier: automatically verify correctness of candidate µGraphs.**  
A discovered µGraph may be fast, but it must be semantically equivalent to the input program. Mirage verifies equivalence through randomized testing over finite fields. By exploiting the mathematical structure of LAX programs, it extends polynomial-identity-testing-style reasoning to provide strong probabilistic correctness guarantees. This avoids floating-point errors and makes the probability of accepting an incorrect µGraph arbitrarily small.

**µGraph Optimizer: further optimize performance after verification.**  
After a µGraph has been verified, Mirage further optimizes it through tensor layout selection, operator scheduling, and memory planning. In this pipeline, the µGraph generator discovers correct and promising implementations, the verifier ensures semantic equivalence, and the optimizer maps the verified implementation into a high-performance GPU program.

Mirage is therefore not merely a kernel fusion system or a schedule optimizer. Its key contribution is to turn cross-level GPU custom-kernel optimization—previously requiring manual expert design—into an automated pipeline that can search, verify, and optimize tensor programs across the GPU hierarchy.

<img width="1098" height="487" alt="Image" src="https://github.com/user-attachments/assets/65f95fa4-a1e0-412e-989d-810ac3f48658" />

<img width="566" height="499" alt="Image" src="https://github.com/user-attachments/assets/adcf9886-0c13-46bd-bbb7-407f31b468db" />
